### PR TITLE
feat(api): add update-stripe-card endpoint

### LINKED ISSUE #54951: Port donate/update-stripe-card changes to new api

### PR DESCRIPTION
**Pull request by ahmaxed**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/freeCodeCamp/issues/54951

### CODE REVIEW

File: api/src/routes/donate.ts [hunk 1]
```diff
@@ -30,6 +30,44 @@ export const donateRoutes: FastifyPluginCallbackTypebox = (
...
+        });
+        return { sessionId: session.id } as const;
+      } catch (error) {
+        fastify.log.error(error);
+        void reply.code(500);
+        return {
+          type: 'danger',
+          message: 'Something went wrong.'
+        } as const;
+      }
```
Comment (ojeytonwilliams) [inline api/src/routes/donate.ts:67]:
```suggestion
    async req => {
      const donation = await fastify.prisma.donation.findFirst({
        where: { userId: req.user?.id }
      });
      if (!donation)
        throw Error(`Stripe donation record not found: ${req.user?.id}`);
      const { customerId, subscriptionId } = donation;
      const session = await stripe.checkout.sessions.create({
        payment_method_types: ['card'],
        mode: 'setup',
        customer: customerId,
        setup_intent_data: {
          metadata: {
            customer_id: customerId,
            subscription_id: subscriptionId
          }
        },
        success_url: `${HOME_LOCATION}/update-stripe-card?session_id={CHECKOUT_SESSION_ID}`,
        cancel_url: `${HOME_LOCATION}/update-stripe-card`
      });
      return { sessionId: session.id } as const;
```
Everything gets a bit simpler if we simply throw any errors. The handler will return `{ message: 'flash.generic-error', type: 'danger' }` and the client handles that the same as any other error response.

File: api/src/schemas/donate/update-stripe-card.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+export const updateStripeCard = {
+  response: {
+    200: Type.Object({
+      sessionId: Type.String()
+    }),
+    default: Type.Object({
+      message: Type.Literal('Something went wrong.'),
+      type: Type.Literal('danger')
+    })
+  }
```
Comment (ojeytonwilliams) [inline api/src/schemas/donate/update-stripe-card.ts:12]:
```suggestion
    default: genericError
```

File: api/src/schemas/donate/update-stripe-card.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Type } from '@fastify/type-provider-typebox';
```
Comment (ojeytonwilliams) [inline api/src/schemas/donate/update-stripe-card.ts:1]:
```suggestion
import { Type } from '@fastify/type-provider-typebox';
import { genericError } from '../types';
```

File: api/src/routes/donate.ts [hunk 4]
```diff
@@ -30,6 +30,44 @@ export const donateRoutes: FastifyPluginCallbackTypebox = (
...
 
+  fastify.put(
+    '/donate/update-stripe-card',
+    {
+      schema: schemas.updateStripeCard
+    },
+    async (req, reply) => {
+      try {
+        const donation = await fastify.prisma.donation.findFirst({
+          where: { userId: req.user?.id }
```
Comment (ojeytonwilliams) [inline api/src/routes/donate.ts:41]: Shouldn't this also check if the `provider` is `stripe`?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api/src/routes/donate.test.ts [hunk 5]
```diff
@@ -302,6 +329,41 @@ describe('Donate', () => {
...
+        });
+        expect(response.body).toEqual({ sessionId: 'checkout_session_id' });
+        expect(response.status).toBe(200);
+
+        expect(true).toBe(true);
+      });
+      it('should return 500 if there is no donation record', async () => {
+        const response = await superPut('/donate/update-stripe-card');
+        expect(response.body).toEqual({
+          message: 'Something went wrong.',
```
Comment (ojeytonwilliams) [inline api/src/routes/donate.test.ts:360]:
```suggestion
          message: 'flash.generic-error',
```
Sorry, missed this.

File: api/src/routes/donate.test.ts [hunk 6]
```diff
@@ -302,6 +329,41 @@ describe('Donate', () => {
...
+              subscription_id: 'sub_test_id'
+            }
+          },
+          success_url:
+            'http://localhost:8000/update-stripe-card?session_id={CHECKOUT_SESSION_ID}'
+        });
+        expect(response.body).toEqual({ sessionId: 'checkout_session_id' });
+        expect(response.status).toBe(200);
+
+        expect(true).toBe(true);
```
Comment (ojeytonwilliams) [inline api/src/routes/donate.test.ts:355]:
```suggestion
```
:thinking:

**Review by ojeytonwilliams — APPROVED** [review]
Looking good! Just the one nit:

**Comment by ahmaxed** [comment]
Added an empty object.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Pull request merged**